keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,534 @@
+- Metrics:
+  - latency
+  - errors
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- E2E test graduates to conformance.
+- Job tracking scales to 10^5 completions.
+
+### Upgrade / Downgrade Strategy
+
+When the feature `JobTrackingWithoutLingeringPods` is enabled for the first
+time, the cluster can have Jobs whose Pods don't have the
+`batch.kubernetes.io/job-completion` finalizer. It would be hard to add the
+finalizer to all Pods while preventing race conditions.
+
+We use `.status.uncountedTerminatedPodUIDs.enabled` to indicate whether the Job
+was created after the feature was enabled. If this boolean is false, the Job
+controller tracks Pods using the legacy algorithm.
+
+The kube-apiserver sets `.status.uncountedTerminatedPodUIDs.enabled = true` when
+the feature gate `JobTrackingWithoutLingeringPods` is enabled, at Job creation.
+
+When the feature is disabled after being enabled for some time, the next time
+the Job controller syncs a Job:
+1. It removes finalizers from all Pods owned by the Job.
+2. Sets `.status.uncountedTerminatedPodUIDs.enabled = false`
+
+### Version Skew Strategy
+
+No implications to node runtime.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: JobTrackingWithoutLingeringPods
+    - Components depending on the feature gate:
+      - kube-apiserver
+      - kube-controller-manager
+  - [ ] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+
+  Yes.
+  
+  - Removing terminated Pods doesn't affect Job status.
+  
+<<[UNRESOLVED: should we count terminating pods?]>>
+  - Pods removed by the user or other controllers count towards failures or
+    completions.
+<<[/UNRESOLVED]>>
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  
+  Yes.
+  The job controller removes finalizers in this case.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+
+  Existing Jobs are tracked with lingering Pods.
+
+* **Are there any tests for feature enablement/disablement?**
+
+  Yes, there are integration tests.
+
+### Rollout, Upgrade and Rollback Planning
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  Try to be as paranoid as possible - e.g., what if some components will restart
+   mid-rollout?
```

Comment:
There's a risk that after you enable the flag and let it run for awhile some of the pods of a long-running job might get removed, which gets reflected in the status appropriately. But after a restart w/o the flag the legacy behaviour will override the numbers based on existing pods. Should the legacy behaviour account for that?

Reply:
The legacy behavior shouldn't change if the feature is disabled (except for removing finalizers)

I added a note for this in "What happens if we reenable the feature if it was previously rolled back?"